kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -474,23 +479,28 @@ public CompletionStage<ResponseFilterResult> forwardResponse(ResponseHeaderData
             return responseFilterResultBuilder().forward(header, response).completed();
         }
 
+        @NonNull
         @Override
-        public <T extends ApiMessage> CompletionStage<T> sendRequest(short apiVersion, ApiMessage request) {
-            short key = request.apiKey();
-            var apiKey = ApiKeys.forId(key);
-            short headerVersion = apiKey.requestHeaderVersion(apiVersion);
-            var header = new RequestHeaderData()
-                    .setCorrelationId(-1)
-                    .setRequestApiKey(key)
-                    .setRequestApiVersion(apiVersion);
-            if (headerVersion > 1) {
-                header.setClientId(filter.getClass().getSimpleName() + "@" + System.identityHashCode(filter));
-            }
-            boolean hasResponse = apiKey != ApiKeys.PRODUCE
-                    || ((ProduceRequestData) request).acks() != 0;
-            var filterPromise = new InternalCompletableFuture<T>(ctx.executor());
+        public <T extends ApiMessage> CompletionStage<T> sendRequest(short apiVersion, @NonNull ApiMessage request) {
+            return sendRequest(new RequestHeaderData().setRequestApiVersion(apiVersion), request)
+                    .thenApply(r -> r == null ? null : (T) r.message());
+        }
+
+        @NonNull
+        @Override
+        public <M extends ApiMessage> CompletionStage<ResponseHeaderAndApiMessage<M>> sendRequest(@NonNull RequestHeaderData header,
```

Comment:
I wonder if we should offer a `RequestEnvelope` type here as well.

Reply:
Let's take this to a discussion

Reply:
https://github.com/kroxylicious/design/discussions/40

Reply:
In the discussion, I'm advocating for us do to envelope. Rob's in a agreed, so I've included a change to this PR to remove the envelope from the sendRequest response.